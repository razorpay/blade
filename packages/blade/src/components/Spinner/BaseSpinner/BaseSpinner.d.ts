import type { IconProps } from '../../Icons';

export { default } from './BaseSpinner.web';

export type BaseSpinnerProps = {
  color: IconProps['color'];
  size: IconProps['size'];
};
