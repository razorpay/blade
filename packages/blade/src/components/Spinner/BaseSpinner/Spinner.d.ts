import type { DotNotationColorStringToken } from '../../../_helpers/types';
import type { Theme } from '../../BladeProvider';
import type { SpinnerSizes } from './spinnerTokens';

export { default } from './Spinner.web';

export type SpinnerProps = {
  color: DotNotationColorStringToken<Theme['colors']>;
  size: keyof SpinnerSizes;
};
