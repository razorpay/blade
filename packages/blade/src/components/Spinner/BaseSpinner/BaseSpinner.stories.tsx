import type { ComponentStory, Meta } from '@storybook/react';
import { Title, Subtitle, Primary, ArgsTable, Stories, PRIMARY_STORY } from '@storybook/addon-docs';
import { Link } from '@storybook/design-system';
import type { ReactElement } from 'react';
import type { BaseSpinnerProps } from './BaseSpinner';
import { BaseSpinner as BaseSpinnerComponent } from './BaseSpinner';
import { getStyledPropsArgTypes } from '~components/Box/BaseBox/storybookArgTypes';

const Page = (): ReactElement => {
  return (
    <>
      <Title />
      <Subtitle>This is the BaseSpinner Internal component.</Subtitle>
      <Link
        withArrow={true}
        href="https://www.figma.com/file/jubmQL9Z8V7881ayUD95ps/Blade---Payment-Light?node-id=14825%3A203592"
        target="_blank"
        rel="noreferrer noopener"
      >
        View in Figma
      </Link>
      <br />
      <br />
      <Title>Example</Title>
      <Subtitle>
        This is the default BaseSpinner. You can change the properties of this spinner using the
        controls in the table below.
      </Subtitle>
      <Primary />
      <Title>Properties</Title>
      <ArgsTable story={PRIMARY_STORY} />
      <Stories />
    </>
  );
};

export default {
  title: 'Components/Spinner/BaseSpinner (Internal)',
  component: BaseSpinnerComponent,
  parameters: {
    docs: {
      page: Page,
    },
  },
  argTypes: getStyledPropsArgTypes(),
} as Meta<BaseSpinnerProps>;

const BaseSpinnerTemplate: ComponentStory<typeof BaseSpinnerComponent> = ({ ...args }) => {
  return <BaseSpinnerComponent {...args} />;
};

export const BaseSpinner = BaseSpinnerTemplate.bind({});
BaseSpinner.storyName = 'Default';
