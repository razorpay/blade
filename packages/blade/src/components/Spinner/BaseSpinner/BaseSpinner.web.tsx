/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-base-to-string */
import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import makeMotionTime from '../../../utils/makeMotionTime';
import BaseLoader from './BaseLoader';
import type { BaseSpinnerProps } from './BaseSpinner.d';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

const AnimatedBaseSpinner = styled.div(
  ({ theme }) => css`
    line-height: 0;
    animation: ${rotate} ${makeMotionTime(theme.motion.duration['2xgentle'])}
      ${theme.motion.easing.standard.effective as string} infinite;
  `,
);

const BaseSpinner = ({ color, size, ...props }: BaseSpinnerProps): React.ReactElement => {
  return (
    <AnimatedBaseSpinner {...props}>
      <BaseLoader color={color} size={size} />
    </AnimatedBaseSpinner>
  );
};

export default BaseSpinner;
