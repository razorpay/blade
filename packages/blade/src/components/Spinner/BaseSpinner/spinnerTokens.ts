import type { DurationString, EasingString } from '~tokens/global/motion';

const dimensions = {
  small: 16,
  medium: 20,
  large: 24,
} as const;

const motion: {
  duration: DurationString;
  easing: EasingString;
} = {
  duration: 'duration.2xgentle',
  easing: 'easing.standard.attentive',
};

export { dimensions, motion };
