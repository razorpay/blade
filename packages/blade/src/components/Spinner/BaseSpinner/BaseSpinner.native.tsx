import React from 'react';
import type { EasingFunctionFactory } from 'react-native-reanimated';
import Animated, {
  cancelAnimation,
  useAnimatedStyle,
  useSharedValue,
  withRepeat,
  withTiming,
} from 'react-native-reanimated';
import BaseLoader from './BaseLoader';
import type { BaseSpinnerProps } from './BaseSpinner.d';
import { motion } from './spinnerTokens';

type WithStyle = {
  style: Record<string, unknown>;
};
type AnimatedBaseSpinnerProps = {
  children: React.ReactNode;
} & WithStyle;

const AnimatedBaseSpinner = (props: AnimatedBaseSpinnerProps): React.ReactElement => {
  const { duration, easing } = motion;
  const rotation = useSharedValue(0);
  const animatedStyles = useAnimatedStyle(() => {
    return {
      transform: [
        {
          rotateZ: `${rotation.value}deg`,
        },
      ],
    };
  }, [rotation.value]);

  React.useEffect(() => {
    rotation.value = withRepeat(
      withTiming(360, {
        duration,
        easing: easing as EasingFunctionFactory,
      }),
      -1,
    );
    return (): void => {
      cancelAnimation(rotation);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return <Animated.View {...props} style={[animatedStyles, props.style]} />;
};

const BaseSpinner = ({
  color,
  size,
  ...props
}: BaseSpinnerProps & WithStyle): React.ReactElement => {
  return (
    <AnimatedBaseSpinner {...props}>
      <BaseLoader color={color} size={size} />
    </AnimatedBaseSpinner>
  );
};

export default BaseSpinner;
