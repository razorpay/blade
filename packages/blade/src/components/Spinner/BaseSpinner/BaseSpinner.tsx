import React from 'react';
import { dimensions } from './spinnerTokens';
import SpinnerIcon from './SpinnerIcon';
import { SpinningBox } from './SpinningBox';
import getIn from '~utils/lodashButBetter/get';
import type { Theme } from '~components/BladeProvider';
import { useTheme } from '~components/BladeProvider';
import { getStyledProps } from '~components/Box/styledProps';
import type { StyledPropsBlade } from '~components/Box/styledProps';
import { metaAttribute, MetaConstants } from '~utils/metaAttribute';
import type { FeedbackColors } from '~tokens/theme/theme';
import BaseBox from '~components/Box/BaseBox';
import { Text } from '~components/Typography';
import type { TestID } from '~utils/types';
import { makeSize } from '~utils/makeSize';
import { makeAccessible } from '~utils/makeAccessible';

type BaseSpinnerProps = {
  /**
   * Sets the color of the spinner.
   *
   * @default 'default'
   */
  color?: 'primary' | 'white' | FeedbackColors;
  /**
   * Sets the label of the spinner.
   *
   * @default 'right'
   */
  label?: string;
  /**
   * Sets the label of the spinner.
   *
   */
  labelPosition?: 'right' | 'bottom';
  /**
   * Sets the size of the spinner.
   *
   * @default 'medium'
   */
  size?: 'medium' | 'large' | 'xlarge';
  /**
   * Sets the aria-label for web & accessibilityLabel react-native.
   *
   */
  accessibilityLabel: string;
} & TestID &
  StyledPropsBlade;

const getColor = ({ color, theme }: { color: BaseSpinnerProps['color']; theme: Theme }): string => {
  if (color && color === 'white') {
    return getIn(theme.colors, 'surface.background.gray.intense');
  }
  if (color && color !== 'primary') {
    return getIn(theme.colors, `feedback.background.${color}.intense`);
  }
  return getIn(theme.colors, 'surface.background.primary.intense');
};

const BaseSpinner = ({
  label,
  labelPosition = 'right',
  accessibilityLabel,
  color = 'neutral',
  size = 'medium',
  testID,
  ...styledProps
}: BaseSpinnerProps): React.ReactElement => {
  const { theme } = useTheme();
  return (
    <BaseBox
      {...metaAttribute({ name: MetaConstants.Spinner, testID })}
      {...getStyledProps(styledProps)}
    >
      <BaseBox
        display="flex"
        alignItems="center"
        flexDirection={labelPosition === 'right' ? 'row' : 'column'}
        {...makeAccessible({
          label: accessibilityLabel,
          role: 'progressbar',
        })}
      >
        <SpinningBox>
          <SpinnerIcon dimensions={makeSize(dimensions[size])} color={getColor({ color, theme })} />
        </SpinningBox>
        {label && label.trim().length > 0 ? (
          <BaseBox
            marginLeft={labelPosition === 'right' ? 'spacing.3' : 'spacing.0'}
            marginTop={labelPosition === 'bottom' ? 'spacing.3' : 'spacing.0'}
          >
            <Text variant="body" weight="regular" size="small" color="surface.text.gray.muted">
              {label}
            </Text>
          </BaseBox>
        ) : null}
      </BaseBox>
    </BaseBox>
  );
};

export type { BaseSpinnerProps };
export { BaseSpinner };
