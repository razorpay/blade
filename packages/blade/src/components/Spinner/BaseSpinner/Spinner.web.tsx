/* eslint-disable @typescript-eslint/restrict-template-expressions */
/* eslint-disable @typescript-eslint/no-base-to-string */
import React from 'react';
import styled, { css, keyframes } from 'styled-components';
import getIn from '../../../utils/getIn';
import makeMotionTime from '../../../utils/makeMotionTime';
import { useTheme } from '../../BladeProvider';
import SpinnerIcon from './SpinnerIcon';
import type { SpinnerProps } from './Spinner.d';
import { getSpinnerSize, motion } from './spinnerTokens';

const rotate = keyframes`
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
`;

const AnimatedSpinner = styled.div(
  () => css`
    line-height: 0;
    animation: ${rotate} ${makeMotionTime(motion.duration)} ${motion.easing as string} infinite;
  `,
);

const Spinner = ({ color, size, ...props }: SpinnerProps): React.ReactElement => {
  const { theme } = useTheme();
  const spinnerSize = getSpinnerSize(size);
  const spinnerColor = getIn(theme.colors, color);

  return (
    <AnimatedSpinner {...props}>
      <SpinnerIcon color={spinnerColor} size={spinnerSize} />
    </AnimatedSpinner>
  );
};

export default Spinner;
