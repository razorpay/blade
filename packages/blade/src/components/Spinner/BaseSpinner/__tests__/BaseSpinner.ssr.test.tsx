import { BaseSpinner } from '../BaseSpinner';
import renderWithSSR from '~src/_helpers/testing/renderWithSSR.web';

describe('<BaseSpinner />', () => {
  it('should render low contrast BaseSpinner with right label', () => {
    const { container, getByText } = renderWithSSR(
      <BaseSpinner
        accessibilityLabel="Loading"
        contrast="low"
        label="Loading"
        labelPosition="right"
      />,
    );
    expect(getByText('Loading')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
