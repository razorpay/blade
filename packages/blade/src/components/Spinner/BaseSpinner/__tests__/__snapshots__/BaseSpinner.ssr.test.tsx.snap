// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BaseSpinner /> should render low contrast BaseSpinner with right label 1`] = `"<div id=\\"root\\" data-reactroot=\\"\\"><div display=\\"flex\\" data-blade-component=\\"spinner\\" class=\\"BaseBoxweb__BaseBox-sc-1icfu8j-0 ivikxH\\"><div display=\\"flex\\" aria-label=\\"Loading\\" role=\\"progressbar\\" class=\\"BaseBoxweb__BaseBox-sc-1icfu8j-0 ePjSmf\\"><div class=\\"BaseBoxweb__BaseBox-sc-1icfu8j-0 SpinningBoxweb__SpinningBox-zswieh-0  lkFZfN\\"><svg aria-hidden=\\"true\\" data-blade-component=\\"icon\\" height=\\"16px\\" viewBox=\\"0 0 24 24\\" width=\\"16px\\" fill=\\"none\\" class=\\"Svgweb__StyledSvg-vcmjs8-0\\"><path d=\\"M24 12C24 18.6274 18.6274 24 12 24C5.37258 24 0 18.6274 0 12C0 5.37258 5.37258 0 12 0C18.6274 0 24 5.37258 24 12ZM3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3C7.02944 3 3 7.02944 3 12Z\\" fill=\\"hsla(217, 18%, 45%, 1)\\" fill-opacity=\\"0.2\\"></path><path d=\\"M24 12C24 13.8937 23.5518 15.7606 22.6921 17.4479C21.8324 19.1352 20.5855 20.5951 19.0534 21.7082C17.5214 22.8213 15.7476 23.556 13.8772 23.8523C12.0068 24.1485 10.0928 23.9979 8.29181 23.4127L9.21886 20.5595C10.5696 20.9984 12.0051 21.1114 13.4079 20.8892C14.8107 20.667 16.141 20.116 17.2901 19.2812C18.4391 18.4463 19.3743 17.3514 20.0191 16.0859C20.6639 14.8204 21 13.4203 21 12H24Z\\" fill=\\"hsla(217, 18%, 45%, 1)\\"></path><path d=\\"M-1.33514e-05 12C-1.33514e-05 10.1063 0.448176 8.23944 1.30791 6.55211C2.16764 4.86479 3.41451 3.4049 4.94656 2.2918C6.47862 1.17869 8.25236 0.443983 10.1228 0.147739C11.9932 -0.148504 13.9072 0.00212896 15.7082 0.587322L14.7811 3.44049C13.4304 3.0016 11.9949 2.88862 10.5921 3.11081C9.18927 3.33299 7.85896 3.88402 6.70992 4.71885C5.56088 5.55367 4.62573 6.64859 3.98093 7.91409C3.33613 9.17958 2.99999 10.5797 2.99999 12H-1.33514e-05Z\\" fill=\\"hsla(217, 18%, 45%, 1)\\"></path></svg></div><div class=\\"BaseBoxweb__BaseBox-sc-1icfu8j-0 jRPjtT\\"><p color=\\"surface.text.subdued.lowContrast\\" font-family=\\"text\\" font-size=\\"75\\" font-weight=\\"regular\\" font-style=\\"normal\\" class=\\"BaseTextweb__StyledBaseText-sc-1ob51ew-0 cShHSz Text__StyledText-wkh56f-0\\" data-blade-component=\\"text\\">Loading</p></div></div></div></div>"`;

exports[`<BaseSpinner /> should render low contrast BaseSpinner with right label 2`] = `
<div
  data-reactroot=""
  id="root"
>
  <div
    class="BaseBoxweb__BaseBox-sc-1icfu8j-0 ivikxH"
    data-blade-component="spinner"
    display="flex"
  >
    <div
      aria-label="Loading"
      class="BaseBoxweb__BaseBox-sc-1icfu8j-0 ePjSmf"
      display="flex"
      role="progressbar"
    >
      <div
        class="BaseBoxweb__BaseBox-sc-1icfu8j-0 SpinningBoxweb__SpinningBox-zswieh-0  lkFZfN"
      >
        <svg
          aria-hidden="true"
          class="Svgweb__StyledSvg-vcmjs8-0"
          data-blade-component="icon"
          fill="none"
          height="16px"
          viewBox="0 0 24 24"
          width="16px"
        >
          <path
            d="M24 12C24 18.6274 18.6274 24 12 24C5.37258 24 0 18.6274 0 12C0 5.37258 5.37258 0 12 0C18.6274 0 24 5.37258 24 12ZM3 12C3 16.9706 7.02944 21 12 21C16.9706 21 21 16.9706 21 12C21 7.02944 16.9706 3 12 3C7.02944 3 3 7.02944 3 12Z"
            fill="hsla(217, 18%, 45%, 1)"
            fill-opacity="0.2"
          />
          <path
            d="M24 12C24 13.8937 23.5518 15.7606 22.6921 17.4479C21.8324 19.1352 20.5855 20.5951 19.0534 21.7082C17.5214 22.8213 15.7476 23.556 13.8772 23.8523C12.0068 24.1485 10.0928 23.9979 8.29181 23.4127L9.21886 20.5595C10.5696 20.9984 12.0051 21.1114 13.4079 20.8892C14.8107 20.667 16.141 20.116 17.2901 19.2812C18.4391 18.4463 19.3743 17.3514 20.0191 16.0859C20.6639 14.8204 21 13.4203 21 12H24Z"
            fill="hsla(217, 18%, 45%, 1)"
          />
          <path
            d="M-1.33514e-05 12C-1.33514e-05 10.1063 0.448176 8.23944 1.30791 6.55211C2.16764 4.86479 3.41451 3.4049 4.94656 2.2918C6.47862 1.17869 8.25236 0.443983 10.1228 0.147739C11.9932 -0.148504 13.9072 0.00212896 15.7082 0.587322L14.7811 3.44049C13.4304 3.0016 11.9949 2.88862 10.5921 3.11081C9.18927 3.33299 7.85896 3.88402 6.70992 4.71885C5.56088 5.55367 4.62573 6.64859 3.98093 7.91409C3.33613 9.17958 2.99999 10.5797 2.99999 12H-1.33514e-05Z"
            fill="hsla(217, 18%, 45%, 1)"
          />
        </svg>
      </div>
      <div
        class="BaseBoxweb__BaseBox-sc-1icfu8j-0 jRPjtT"
      >
        <p
          class="BaseTextweb__StyledBaseText-sc-1ob51ew-0 cShHSz Text__StyledText-wkh56f-0"
          color="surface.text.subdued.lowContrast"
          data-blade-component="text"
          font-family="text"
          font-size="75"
          font-style="normal"
          font-weight="regular"
        >
          Loading
        </p>
      </div>
    </div>
  </div>
</div>
`;
